cpp/src/io/json/nested_json.hpp

```diff
@@ -0,0 +1,268 @@
+  ListBegin,
+  /// End-of-list token (on encounter of semantic ']')
+  ListEnd,
+  /// Beginning-of-field-name token (on encounter of first quote)
+  FieldNameBegin,
+  /// End-of-field-name token (on encounter of a field name's second quote)
+  FieldNameEnd,
+  /// Beginning-of-string-value token (on encounter of the string's first quote)
+  StringBegin,
+  /// End-of-string token (on encounter of a string's second quote)
+  StringEnd,
+  /// Beginning-of-value token (first character of literal or numeric)
+  ValueBegin,
+  /// Post-value token (first character after a literal or numeric string)
+  ValueEnd,
+  /// Beginning-of-error token (on first encounter of a parsing error)
+  ErrorBegin,
+  /// Total number of tokens
+  NUM_TOKENS
+};
+
+namespace detail {
+/**
+ * @brief Identifies the stack context for each character from a JSON input. Specifically, we
+ * identify brackets and braces outside of quoted fields (e.g., field names, strings).
+ * At this stage, we do not perform bracket matching, i.e., we do not verify whether a closing
+ * bracket would actually pop a the corresponding opening brace.
+ *
+ * @param[in] d_json_in The string of input characters
+ * @param[out] d_top_of_stack Will be populated with what-is-on-top-of-the-stack for any given input
+ * character of \p d_json_in, where a '{' represents that the corresponding input character is
+ * within the context of a struct, a '[' represents that it is within the context of an array, and a
+ * '_' symbol that it is at the root of the JSON.
+ * @param[in] stream The cuda stream to dispatch GPU kernels to
+ */
+void get_stack_context(device_span<SymbolT const> d_json_in,
+                       SymbolT* d_top_of_stack,
+                       rmm::cuda_stream_view stream);
+
+/**
+ * @brief Parses the given JSON string and emits a sequence of tokens that demarcate relevant
+ * sections from the input.
+ *
+ * @param[in] d_json_in The JSON input
+ * @param[out] d_tokens_out Device memory to which the parsed tokens are written
+ * @param[out] d_tokens_indices Device memory to which the indices are written, where each index
+ * represents the offset within \p d_json_in that cause the input being written
+ * @param[out] d_num_written_tokens The total number of tokens that were parsed
+ * @param[in] stream The CUDA stream to which kernels are dispatched
+ */
+void get_token_stream(device_span<SymbolT const> d_json_in,
+                      PdaTokenT* d_tokens,
+                      SymbolOffsetT* d_tokens_indices,
+                      SymbolOffsetT* d_num_written_tokens,
+                      rmm::cuda_stream_view stream);
+
+/**
+ * @brief Parses the given JSON string and generates a tree representation of the given input.
+ *
+ * @param input The JSON input in host memory
+ * @param d_input The JSON input in device memory
+ * @param stream The CUDA stream to which kernels are dispatched
+ * @return
+ */
+json_column get_json_columns(host_span<SymbolT const> input,
```

Comment:
This function is only ever used in `nested_json_gpu.cu`.  Seems like we should just get rid of this declaration and rearrange the order of the functions in .cu file.   

Are there any other declarations in here that can be moved into `nested_json_gpu.cu`?  I see a few other functions that look internal-ish but are actually called by tests.

Reply:
Yeah, motivation for exposing `get_stack_context` and `get_token_stream` being two-fold: 
- As @karthikeyann mentioned, we want to have them unit tested.
- They constitute potentially reusable components, in particular tokenizer `get_token_stream`